### DISCUSSION TITLE
Failed to install wheel in Ubuntu16.04

### DISCUSSION THREAD

**Discussion by rise0chen** [PyO3/maturin#1368] [Q&A] [answered]
## Build
```bash
maturin build
```
```bash
Built wheel for abi3 Python ≥ 3.7 to ***-cp37-abi3-manylinux_2_24_x86_64.whl`
````

## Install
```bash
pip install ***-cp37-abi3-manylinux_2_24_x86_64.whl
```
```bash
ERROR: ***-cp37-abi3-manylinux_2_24_x86_64.whl is not a supported wheel on this platform.
```

## 
I think, it should build `***-cp37-abi3-manylinux_2_17_aarch64.manylinux2014_aarch64.whl`.

It can work if I rename `***-cp37-abi3-manylinux_2_24_x86_64.whl` to `***-cp37-abi3-manylinux_2_17_aarch64.manylinux2014_aarch64.whl`.

```bash
maturin build --compatibility manylinux2014
```
```bash
maturin failed
  Caused by: Error ensuring manylinux_2_17 compliance
  Caused by: Your library is not manylinux_2_17 (aka manylinux2014) compliant because of the presence of too-recent versioned symbols: ["libc.so.6 offending symbols: __cxa_thread_atexit_impl@@GLIBC_2.18"]. Consider building in a manylinux docker container
```
GLIBC version is `2.23-0ubuntu11.3`

### Your maturin version (`maturin --version`)

0.14.7

### Your Python version (`python -V`)

3.7.4

### Your pip version (`pip -V`)

22.3.1

### What bindings you're using

pyo3

### Does `cargo build` work?

- [X] Yes, it works

### If on windows, have you checked that you aren't accidentally using unix path (those with the forward slash `/`)?

- [ ] Yes

### Steps to Reproduce

(.env) root@f0fca0e971de:~/sdk# RUST_LOG=maturin=debug maturin build  --features ffi_py,mdns
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.7
🐍 Not using a specific python interpreter
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
📦 Built wheel for abi3 Python ≥ 3.7 to /root/sdk/target/wheels/lebai_sdk-0.1.1-cp37-abi3-manylinux_2_24_x86_64.whl

**Comment by messense** [comment, marked as answer]
You need to ensure your build environment is manylinux compliant not the other way around. If you want manylinux2014, build your wheels in a manylinux2014 docker container.

See https://github.com/pypa/manylinux

**Reply by messense** [reply]
> GLIBC version is `2.23-0ubuntu11.3`

And this is obviously not manylinux2014 (aka manylinux_2_17, that is GLIBC `2.17`) compliant.

**Comment by rise0chen** [comment]
Why it can work when I rename `***-cp37-abi3-manylinux_2_24_x86_64.whl` to `***-cp37-abi3-manylinux_2_17_aarch64.manylinux2014_aarch64.whl`.

**Reply by messense** [reply]
Well, you are running on a GLIBC 2.23 system so of course it will work since glibc 2.23 is backward compatible with softwares built with glibc 2.17 because of symbol versioning.

It won't work if you run it on a CentOS 7 system which defaults to glibc 2.17 since your package requires at least glibc 2.18 because of the `__cxa_thread_atexit_impl@@GLIBC_2.18` symbol.

When you name it `manylinux2014` you are making a promise that it will work on any Linux system that has glibc 2.17 and later. Just rename the file doesn't uphold this promise, but if you don't care about older systems, it's fine to use `--skip-auditwheel` **as long as you don't dynamic link to other forbidden shared libraries**. I suggest you read [PEP 513](https://peps.python.org/pep-0513/) if you intend to do so.

**Reply by rise0chen** [reply]
Thinks, `--skip-auditwheel` is useful.
abandon glibc 2.17 to 2.22 is acceptable for me.

**Comment by konstin** [comment]
What you can also try instead of the manylinux container is to install `maturin[zig]` and build with `--zig`. Using zig, you should be able to build for arbitrary glibc versions on arbitrary host platforms
